build/lib/src/asset/id.dart

```diff
@@ -154,3 +151,12 @@ String _normalizePath(String path) {
   // Collapse "." and "..".
   return p.posix.normalize(path);
 }
+
+Uri _constructUri(AssetId id) {
+  final originalSegments = p.split(id.path);
+  final isLib = originalSegments.first == 'lib';
+  final scheme = isLib ? 'package' : 'asset';
+  final pathSegments = isLib ? originalSegments.skip(1) : originalSegments;
+  return new Uri(
+      scheme: scheme, pathSegments: [id.package]..addAll(pathSegments));
```

Comment:
so `pathSegments` escapes and `path` doesn't?

Reply:
`path` escapes some stuff, but not everything. It leaves delimiters, even when they are invalid.
https://github.com/dart-lang/sdk/blob/80a571d5460bbc734e709d7437af2c7df11e582a/sdk/lib/core/uri.dart#L85

`pathSegments` escapes everything which is, I think, what we want.